src/container/amlManager.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+"use strict";
+import { AzureMachineLearningWorkspaces } from "azure-arm-machinelearningservices";
+import { Workspace } from "azure-arm-machinelearningservices/lib/models";
+import { Workspaces } from "azure-arm-machinelearningservices/lib/operations";
+import { HttpOperationResponse, ServiceClient, ServiceClientCredentials } from "ms-rest";
+import * as vscode from "vscode";
+import { Constants } from "../common/constants";
+import { UserCancelledError } from "../common/UserCancelledError";
+import { Utility } from "../common/utility";
+import { AzureAccount, AzureSession } from "../typings/azure-account.api";
+import { AmlWorkspaceQuickPickItem } from "./models/amlWorkspaceQuickPickItem";
+
+export class AmlManager {
+    private readonly azureAccount: AzureAccount;
+    constructor() {
+        this.azureAccount = vscode.extensions.getExtension<AzureAccount>("ms-vscode.azure-account")!.exports;
+    }
+
+    public async selectAmlImage(): Promise<string> {
+        const workspaceItem: AmlWorkspaceQuickPickItem = await this.selectWorkspace();
+        if (workspaceItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        const imageItem: vscode.QuickPickItem = await this.selectImage(workspaceItem.workspace, workspaceItem.azureSubscription.session);
+        if (imageItem === undefined) {
+            throw new UserCancelledError();
+        }
+
+        return imageItem.description;
+    }
+
+    private async selectWorkspace(): Promise<AmlWorkspaceQuickPickItem> {
+        await Utility.waitForAzLogin(this.azureAccount);
+
+        const workspaceItem: AmlWorkspaceQuickPickItem =
+            await vscode.window.showQuickPick(this.loadWorkspaceItems(), { placeHolder: "Select Azure Machine Learning Workspace", ignoreFocusOut: true });
+
+        return workspaceItem;
+    }
+
+    private async loadWorkspaceItems(): Promise<AmlWorkspaceQuickPickItem[]> {
+        try {
+            await this.azureAccount.waitForFilters();
+            const workspacePromises: Array<Promise<AmlWorkspaceQuickPickItem[]>> = [];
+            for (const azureSubscription of this.azureAccount.filters) {
+                const client: Workspaces = new AzureMachineLearningWorkspaces(
+                    azureSubscription.session.credentials,
+                    azureSubscription.subscription.subscriptionId!,
```

Comment:
! is for?

Reply:
Just a hint to tell the compiler that the `subscriptionId` cannot be null, which can be useful with some strict compiler flag. See https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#non-null-assertion-operator. This is copied from the sample code https://github.com/Microsoft/vscode-azure-account/blob/master/sample/src/extension.ts#L78